File: test/conformance/read_surface_test.rb
# frozen_string_literal: true

require_relative 'conformance_helper'

# The addresses Rails still answers that are not the upload (#291): the
# redirects, the retirements, the catch-all, the availability feed and the
# sitemap. #302 moves the redirects into nginx and #303 the sitemap into the
# static build; these hold whichever side answers, which is the point.
#
# Ported from redirects_test and retired_routes_test, where they talked to the
# app in-process. Paths only in the Location comparisons: the application
# answers relative or absolute depending on who built the redirect, and a
# browser treats the two the same.
class ReadSurfaceConformanceTest < Conformance::Case
  surface :read

  PERMANENT = {
    '/home' => '/', '/introduction' => '/', '/aaa' => '/', '/fpv' => '/low-latency',
    '/our-projects' => '/ecosystem', '/our-software' => '/ecosystem', '/our-channels' => '/community',
    '/support-open-source' => '/donate', '/sponsor' => '/donate',
    '/hardware' => '/supported-hardware/featured', '/SDK' => '/supported-hardware',
    '/supported-hardware' => '/supported-hardware/featured'
  }.freeze

  GONE = ['/binaries', '/binaries.json', '/telemetry', '/telemetry/anything', '/merchandise',
          '/ru/merchandise', '/zh/merchandise', '/snapshots/12345'].freeze

  GITHUB = %w[coupler firmware ipctool microbe-web smolrtsp yaml-cli wiki].freeze

  def location_path(response)
    URI(response['Location']).path
  end

  test 'the old addresses are permanent redirects to their pages' do
    wrong = PERMANENT.filter_map do |from, to|
      response = get(from)
      next if response.code == '301' && location_path(response) == to

      "#{from}: want 301 #{to}, got #{response.code} #{response['Location']}"
    end

    assert_empty wrong
  end

  # 302, not 301: /about is meant to become a page of its own, and a browser
  # caches a 301 indefinitely.
  test '/about is a temporary redirect to /community' do
    response = get('/about')

    assert_equal '302', response.code
    assert_equal '/community', location_path(response)
  end

  test 'what was retired says it is gone' do
    wrong = GONE.filter_map do |path|
      code = get(path).code
      "#{path}: #{code}" unless code == '410'
    end

    assert_empty wrong, 'a retired address that answers anything but 410 tells a crawler it moved'
  end

  # The camera's newest frame as a file, retired in #235. nginx refuses it
  # before Rails hears of it; Rails alone would run the catch-all. So this one
  # holds only through the vhost, and is the example of a behaviour a port
  # does not have to carry because it is not the application's.
  test 'through nginx, a camera frame by URL is gone' do
    skip 'the vhost answers this, not the application' if Conformance.direct?

    assert_equal '410', get('/open-wall/camera/0123456789abcdef.jpg').code
  end

  test 'every GitHub shortcut points at its OpenIPC repository' do
    wrong = GITHUB.product(['', '/some/deep/path']).filter_map do |repo, suffix|
      response = get("/#{repo}#{suffix}")
      next if response.code.start_with?('3') &&
              response['Location'].to_s.match?(%r{\Ahttps://github\.com/OpenIPC/#{Regexp.escape(repo)}/?\z})

      "/#{repo}#{suffix}: #{response.code} #{response['Location']}"
    end

    assert_empty wrong
  end

  test 'an unknown address goes to the home page, and sets nothing' do
    response = get('/no-such-page-at-all', headers: { 'Referer' => 'https://example.test/private' })

    assert_equal '302', response.code
    assert_equal '/', location_path(response)
  end

  # What 378 prerendered pages read on load to say what a visitor can do with
  # each chip (#162). The states are the page's vocabulary; a fourth one is a
  # page that renders nothing for it.
  test 'the availability feed names every chip with a state the pages know' do
    response = get('/api/v1/hardware/availability.json')

    assert_equal '200', response.code
    assert_match %r{\Aapplication/json}, response['Content-Type']
    body = JSON.parse(response.body)
    assert_kind_of String, body['generated_at']
    assert_operator body['socs'].size, :>, 100, 'the feed lost most of the catalogue'
    assert_empty(body['socs'].values.uniq - %w[wizard firmware_only none])
    assert_equal body['socs'].keys.sort, body['socs'].keys, 'the feed is ordered by slug'
    assert_nil response['Set-Cookie']
  end

  test 'the sitemap is XML and offers the catalogue in three languages' do
    response = get('/sitemap.xml')

    assert_equal '200', response.code
    assert_match %r{\A(application|text)/xml}, response['Content-Type']
    assert_includes response.body, '<urlset'
    %w[en ru zh x-default].each do |code|
      assert_includes response.body, %(hreflang="#{code}"), "no #{code} alternates"
    end
    assert_match %r{/cameras/vendors/[a-z0-9._-]+/socs/[a-z0-9._-]+</loc>}, response.body
  end

  test 'the health check answers' do
    assert_equal '200', get('/up').code
  end
end

File: service/conformance/read_surface_test.go
package conformance

import (
	"encoding/json"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"testing"
)

// The addresses that are not the upload or the wall (#291): the redirects,
// the retirements, the catch-all, the availability feed and the sitemap. #302
// moves the redirects into nginx and #303 the sitemap into the static build;
// these hold whichever side answers, which is the point.
//
// Paths only in the Location comparisons: an application answers relative or
// absolute depending on who built the redirect, and a browser treats the two
// the same.

var permanent = map[string]string{
	"/home": "/", "/introduction": "/", "/aaa": "/", "/fpv": "/low-latency",
	"/our-projects": "/ecosystem", "/our-software": "/ecosystem", "/our-channels": "/community",
	"/support-open-source": "/donate", "/sponsor": "/donate",
	"/hardware": "/supported-hardware/featured", "/SDK": "/supported-hardware",
	"/supported-hardware": "/supported-hardware/featured",
}

var gone = []string{"/binaries", "/binaries.json", "/telemetry", "/telemetry/anything", "/merchandise",
	"/ru/merchandise", "/zh/merchandise", "/snapshots/12345"}

var github = []string{"coupler", "firmware", "ipctool", "microbe-web", "smolrtsp", "yaml-cli", "wiki"}

func locationPath(r *response) string {
	u, err := url.Parse(r.Header.Get("Location"))
	if err != nil {
		return ""
	}
	return u.Path
}

func TestTheOldAddressesArePermanentRedirectsToTheirPages(t *testing.T) {
	s := start(t, "read")
	var wrong []string
	for from, to := range permanent {
		r := s.get(from)
		if r.StatusCode != 301 || locationPath(r) != to {
			wrong = append(wrong, from+": want 301 "+to+", got "+r.Status+" "+r.Header.Get("Location"))
		}
	}
	sort.Strings(wrong)
	if len(wrong) > 0 {
		t.Error(strings.Join(wrong, "\n"))
	}
}

// 302, not 301: /about is meant to become a page of its own, and a browser
// caches a 301 indefinitely.
func TestAboutIsATemporaryRedirectToCommunity(t *testing.T) {
	s := start(t, "read")
	r := s.get("/about")
	if r.StatusCode != 302 || locationPath(r) != "/community" {
		t.Errorf("/about: %d %q", r.StatusCode, r.Header.Get("Location"))
	}
}

// A retired address that answers anything but 410 tells a crawler it moved.
func TestWhatWasRetiredSaysItIsGone(t *testing.T) {
	s := start(t, "read")
	for _, path := range gone {
		if r := s.get(path); r.StatusCode != 410 {
			t.Errorf("%s: %d", path, r.StatusCode)
		}
	}
}

// The camera's newest frame as a file, retired in #235. nginx refuses it
// before any application hears of it, so this holds only through the vhost.
func TestThroughNginxACameraFrameByURLIsGone(t *testing.T) {
	s := start(t, "read")
	if direct() {
		t.Skip("the vhost answers this, not the application")
	}
	if r := s.get("/open-wall/camera/0123456789abcdef.jpg"); r.StatusCode != 410 {
		t.Errorf("%d", r.StatusCode)
	}
}

func TestEveryGitHubShortcutPointsAtItsOpenIPCRepository(t *testing.T) {
	s := start(t, "read")
	for _, repo := range github {
		want := regexp.MustCompile(`^https://github\.com/OpenIPC/` + regexp.QuoteMeta(repo) + `/?$`)
		for _, suffix := range []string{"", "/some/deep/path"} {
			r := s.get("/" + repo + suffix)
			if r.StatusCode/100 != 3 || !want.MatchString(r.Header.Get("Location")) {
				t.Errorf("/%s%s: %d %q", repo, suffix, r.StatusCode, r.Header.Get("Location"))
			}
		}
	}
}

func TestAnUnknownAddressGoesToTheHomePageAndSetsNothing(t *testing.T) {
	s := start(t, "read")
	r := s.get("/no-such-page-at-all", map[string]string{"Referer": "https://example.test/private"})
	if r.StatusCode != 302 || locationPath(r) != "/" {
		t.Errorf("%d %q", r.StatusCode, r.Header.Get("Location"))
	}
	if r.Header.Get("Set-Cookie") != "" {
		t.Error("the catch-all set a cookie")
	}
}

// What 378 prerendered pages read on load to say what a visitor can do with
// each chip (#162). The states are the page's vocabulary; a fourth one is a
// page that renders nothing for it.
func TestTheAvailabilityFeedNamesEveryChipWithAStateThePagesKnow(t *testing.T) {
	s := start(t, "read")
	r := s.get("/api/v1/hardware/availability.json")
	if r.StatusCode != 200 || !strings.HasPrefix(r.Header.Get("Content-Type"), "application/json") {
		t.Fatalf("%d %q", r.StatusCode, r.Header.Get("Content-Type"))
	}
	var body struct {
		GeneratedAt *string `json:"generated_at"`
	}
	json.Unmarshal(r.body, &body)
	if body.GeneratedAt == nil {
		t.Error("no generated_at")
	}
	var outer map[string]json.RawMessage
	json.Unmarshal(r.body, &outer)
	slugs := keys(t, outer["socs"])
	var socs map[string]string
	json.Unmarshal(outer["socs"], &socs)
	if len(socs) <= 100 {
		t.Errorf("the feed lost most of the catalogue: %d", len(socs))
	}
	for slug, state := range socs {
		if state != "wizard" && state != "firmware_only" && state != "none" {
			t.Errorf("%s: %q", slug, state)
		}
	}
	if !sort.StringsAreSorted(slugs) {
		t.Error("the feed is not ordered by slug")
	}
	if r.Header.Get("Set-Cookie") != "" {
		t.Error("the feed set a cookie")
	}
}

func TestTheSitemapIsXMLAndOffersTheCatalogueInThreeLanguages(t *testing.T) {
	s := start(t, "read")
	r := s.get("/sitemap.xml")
	if r.StatusCode != 200 || !regexp.MustCompile(`^(application|text)/xml`).MatchString(r.Header.Get("Content-Type")) {
		t.Fatalf("%d %q", r.StatusCode, r.Header.Get("Content-Type"))
	}
	body := string(r.body)
	if !strings.Contains(body, "<urlset") {
		t.Error("no <urlset")
	}
	for _, code := range []string{"en", "ru", "zh", "x-default"} {
		if !strings.Contains(body, `hreflang="`+code+`"`) {
			t.Errorf("no %s alternates", code)
		}
	}
	if !regexp.MustCompile(`/cameras/vendors/[a-z0-9._-]+/socs/[a-z0-9._-]+</loc>`).MatchString(body) {
		t.Error("no catalogue pages")
	}
}

func TestTheHealthCheckAnswers(t *testing.T) {
	s := start(t, "read")
	if r := s.get("/up"); r.StatusCode != 200 {
		t.Errorf("/up: %d", r.StatusCode)
	}
}
